### PR TITLE
Add generator option skip_action_controller

### PR DESCRIPTION
**Pull request by dudo**
### Summary

Allow `rails new --skip_action_controller`. 

### Other Information

This would allow us to create an app with a bit less boilerplate when we don't need to accept web requests. sidekiq / cron come to mind, but I also have plans to use this with Twirp, which doesn't use controllers.

### CODE REVIEW

File: railties/lib/rails/generators/app_base.rb [hunk 1]
```diff
@@ -24,80 +24,83 @@ def self.strict_args_position
...
 
-        class_option :skip_keeps,          type: :boolean, default: false,
-                                           desc: "Skip source control .keep files"
+        class_option :skip_keeps,             type: :boolean, default: false,
+                                              desc: "Skip source control .keep files"
 
-        class_option :skip_action_mailer,  type: :boolean, aliases: "-M",
-                                           default: false,
-                                           desc: "Skip Action Mailer files"
+        class_option :skip_action_controller, type: :boolean, default: false,
```
Comment (dudo) [inline railties/lib/rails/generators/app_base.rb:39]: This is the only new option... needed to add spaces to all the rest 😩

**Comment by rafaelfranca** [comment]
Thank you for the pull request but Rails is a web framework, Action Controller is the only required component of Rails aside of railties and Active Support, so I don't think it makes sense to skip it.

**Comment by dudo** [comment]
@rafaelfranca rack is what makes it a webserver. If paired with something like https://github.com/twitchtv/twirp-ruby, we don't need controllers. 

Easy enough to edit a new rails app, but this would make it cleaner to pair with a `--template`

**Comment by rafaelfranca** [comment]
rack is just an interface. What implements that interface is `Rails::Engine` and Action Dispatch, but even though we can have Rails applications without controller I don't think it is worth supporting skipping generating the controllers.

**Comment by rafaelfranca** [comment]
But, let's say we support this use case, why would you remove Action Controller but not Action View?

**Comment by dudo** [comment]
Action View could definitely go. Probably have an implicit `--api` to make it easy. That would catch the helpers, too.

What I don't understand is how the `delete_app_helpers_if_api_option` methods get called? Seems like we'd want one of those instead of the empty_directory_with_keep_file conditionals? I couldn't figure out for the life of me how those are triggered, though.

**Comment by dudo** [comment]
in app_generator.rb initializer, maybe?

```ruby
@options[:api] = true if @options[:skip_action_controller]
```

**Comment by matthewd** [comment]
If you want Rack without Action Controller, I guess I'm struggling to visualise in what ways that result is going to feel like "Rails Application" and not "Rack Application that uses Active Record" :confused:

Without Rails request handling or routing, it doesn't seem like we'd have much to beyond an `app/models` directory.

**Comment by matthewd** [comment]
> Presumably added to a rails API, so we'd still have controllers

.. but you removed Action Controller?

Action Controller, as part of Action Pack, is an explicit dependency of Railties. Omitting it is a hugely larger change than leaving out `app/controllers/`, and having done so, I'm really not sure what's left.

**Comment by dudo** [comment]
Oh... I'm sorry, I got my projects mixed up. This wouldn't use that gem at all, but rely on handlers called directly from config.ru. Controllers would be replaced by Handlers. 

You're right, though. This would leave us with a shell of Rails, and that's kinda the point. It gives folks more room to tinker, I feel.

**Comment by matthewd** [comment]
I think we'd need a more concrete proposal of what an ActionController-less application would look like (which files still exist, which otherwise-default files are omitted, which files are modified and how?). Outside of generators, what changes are needed in Railties-proper to avoid the Action Controller dependency? Are we assuming there's still a Rack server, just not Action Controller? Do the various `rails` subcommands still work?

**Comment by dudo** [comment]
@matthewd let's see what I can answer:

> Outside of generators, what changes are needed in Railties-proper to avoid the Action Controller dependency? 

Not sure about this one. I can't seem to even find the dependencies you have in mind.

> Are we assuming there's still a Rack server, just not Action Controller? 

This seems like the crux of it. My primary use case for this change is to use something non-rack entirely, in the form of [grpc](https://github.com/grpc/grpc/tree/master/src/ruby). `rails server` inherits from `rack server` which is all about receiving http1 requests. Long term, I'd love to see an option passed to `server` that allowed us to use a custom server... or even a way to define custom commands like `rails rpc`. If anything, maybe this would open the door for better decoupling rails from rack? `ActionHandler`, maybe?

> Do the various rails subcommands still work?

I don't see why not. `server` is the only one I'm unsure of, but [the rest](https://github.com/rails/rails/blob/main/railties/lib/rails/commands.rb) are all relevant.

**Comment by matthewd** [comment]
Okay, that seems pretty squarely outside Rails boundaries. Rails is a web framework.

You can use Active Record migrations etc outside the bounds of a Rails application, and it feels like that's a better fit for your use case.

To revisit this we'd need to be considering a holistic PR that demonstrated a generator option which produced something that didn't sound like a half-broken web application. In its current "just omit Action Controller" state, this is not that. (See e.g. everything in the generated `config/`.)

**Pull request closed without merging**